### PR TITLE
Implement feature: firecrawl caching (Issue 46)

### PR DESCRIPTION
**Pull request by niti-go**
## Summary

This PR implements **automatic caching for Firecrawl web scraping results** to reduce API credit usage and improve performance. The caching is transparent and requires zero manual management from users—it works automatically whenever `<web>` tags are used in prompts. It can optionally be disabled with `export FIRECRAWL_CACHE_ENABLE=false`.   

**Addresses:** Issue #46 in the pdd repository - Cache firecrawl results so it doesn't use up the API credit
Work began in PR #151 ; this branch finishes the full implementation.

### What's Being Added

- **Transparent Automatic Caching**: Web content scraped via `<web>` tags in prompts is automatically cached for 24 hours (configurable)
- **Persistent Storage**: SQLite-based cache at `.pdd/cache/firecrawl.db` persists across sessions
- **Smart URL Normalization**: Handles URL variations (case-insensitive, removes tracking parameters)
- **Configurable Limits**: Control cache TTL, max size, max entries, and auto-cleanup behavior
- **CLI Management Commands**: Debug and inspect cache with `pdd firecrawl-cache` subcommands
- **Zero Configuration**: Works out-of-the-box with sensible defaults

### Environment Variables

Users can customize caching behavior:
```bash
export FIRECRAWL_CACHE_ENABLE=false           # Disable caching (default: true)
export FIRECRAWL_CACHE_TTL_HOURS=48           # Cache for 48 hours (default: 24)
export FIRECRAWL_CACHE_MAX_SIZE_MB=200        # Max cache size in MB (default: 100)
export FIRECRAWL_CACHE_MAX_ENTRIES=2000       # Max entries (default: 1000)
export FIRECRAWL_CACHE_AUTO_CLEANUP=false     # Disable auto cleanup (default: true)
```

### Cache Management Commands

```bash
pdd firecrawl-cache stats              # View cache statistics
pdd firecrawl-cache clear              # Clear all cached entries
pdd firecrawl-cache info               # View cache configuration
pdd firecrawl-cache check <url>        # Check if a URL is cached
```

### Core Implementation
- **`pdd/firecrawl_cache.py`** (437 lines)
  - Main caching module with `FirecrawlCache` class
  - SQLite persistence layer with TTL-based expiration
  - URL normalization logic
  - LRU eviction when cache limits exceeded
  - Stats and cleanup utilities

- **`pdd/commands/firecrawl.py`** (177 lines)
  - New CLI command group: `pdd firecrawl-cache`
  - Subcommands: `stats`, `clear`, `info`, `check`
  - Rich terminal UI for cache visualization

### Integration
- **`pdd/preprocess.py`** (31 lines changed)
  - Integration with web tag processing
  - Automatic cache lookup/storage on web scraping

### Documentation
- **`README.md`** (30 lines added)
  - New "Firecrawl Web Scraping Cache" section
  - Usage instructions and examples
  - Configuration reference
  - Command examples for cache management

### Example Context
- **`context/firecrawl_example.py`** (125 lines changed)
  - Updated example demonstrating cache behavior

### Unit Tests
- **`tests/test_firecrawl_cache.py`** (481 lines)
  - 40+ test cases covering:
    - Cache initialization and persistence
    - URL normalization edge cases
    - TTL expiration logic
    - Size and entry limits with LRU eviction
    - Concurrent access scenarios
    - Environment variable configuration
    - Cache statistics and cleanup

### Integration Tests
- **`tests/test_commands_firecrawl.py`** (737 lines)
  - Command functionality tests for all `pdd firecrawl-cache` subcommands
  - CLI output validation
  - Error handling scenarios

### Preprocess Tests
- **`tests/test_preprocess.py`** (54 lines changed)
  - Updated to work with caching disabled in tests (prevents API calls)
  - Ensures cache doesn't interfere with unit testing

## Key Design Decisions

1. **Singleton Pattern**: Single cache instance per process avoids duplicate database connections
2. **SQLite Backend**: Lightweight, no external dependencies, built-in Python support
3. **URL Normalization**: Consistent caching despite URL variations (tracking params, case differences)
4. **Transparent Operation**: No API changes required; existing prompts automatically benefit
5. **Configurable Limits**: LRU eviction prevents unbounded growth
6. **Environment Variables**: Non-breaking configuration approach

## Testing Notes

- All 40+ cache-specific tests pass
- 10+ command tests pass for CLI management
- Preprocess tests updated to disable caching to avoid API side effects
- Example files validated

## Usage Example

```bash
# First time: fetches from Firecrawl API
pdd generate --output src/module.py prompts/module_python.prompt

# Second time: same URL is served from cache (24h default TTL)
pdd generate --output src/module2.py prompts/module2_python.prompt

# Check what's cached
pdd firecrawl-cache stats

# Clear cache if needed
pdd firecrawl-cache clear
```

### CODE REVIEW

**Comment by niti-go** [comment]
I imagine most users won't need to think about caching and manually manage it. Do you think it makes sense for the cache management commands (such as viewing usage statistics or checking if a website is stored in the cache) to be accessible as the main command`pdd firecrawl-cache`? Or would it be better to simplify it? For example, we could only allow users to enable/disable the cache, but this would reduce customizability.

File: tests/test_commands_firecrawl.py [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+            'active_entries': 1,
+            'expired_entries': 0
+        }
+        mock_confirm.return_value = True
+        mock_clear.return_value = 1
+
+        result = runner.invoke(cli, ["firecrawl-cache", "clear"])
+
+        assert result.exit_code == 0
+        mock_confirm.assert_called_once_with("Clear 1 cached entries?")  # Note: grammar issue in original
```
[the remark this answers is not in the document]
Comment (niti-go) [reply tests/test_commands_firecrawl.py]: The grammar issue is just a UX polish that doesn't affect cache functionality.

File: .env.example [hunk 2]
```diff
@@ -1,18 +1,9 @@
...
-VERTEX_LOCATION=us-central1
-
-# PDD Configuration
-PDD_PATH=/path/to/your/pdd
\ No newline at end of file
+OPENAI_API_KEY =your-openai-api-key
+GOOGLE_API_KEY = your-google-api-key
+ANTHROPIC_API_KEY = your-anthropic-api-key
+PDD_PATH = Where-Your-PDD-Root-Is-At
+DEEKSEEK_API_KEY = your-deekseek-api-key
```
[the remark this answers is not in the document]
Comment (niti-go) [reply .env.example]: I reverted this file to the .env.example currently in main, I didn't mean to modify it.

**Comment by niti-go** [comment]
@gltanaka I reviewed the Copilot comments.

**Comment by gltanaka** [comment]
## Review & Fixes for Firecrawl Caching 🎉

Great work on this PR! The caching implementation is solid — clean SQLite-based design, good TTL/LRU eviction, URL normalization, and comprehensive test coverage (143 tests!). We benchmarked the caching and saw a **306x speedup** on cache hits (0.7ms vs 211ms per call), which is a huge win for API credit savings.

We went ahead and applied a handful of small fixes on top of your work so this can move toward merging faster. Here's what we found and addressed:

### Bugs fixed

1. **Duplicate TTL parsing in `preprocess.py`** — There was a bare `int(os.environ.get('FIRECRAWL_CACHE_TTL_HOURS', 24))` in `process_web_tags()` that duplicated the parsing already handled by `FirecrawlCache._load_config()`. If someone set the env var to a non-numeric value, this would crash. Removed the duplicate and let the cache handle TTL internally.

2. **Unguarded `int()` for env vars in `_load_config()`** — `FIRECRAWL_CACHE_MAX_SIZE_MB` and `FIRECRAWL_CACHE_MAX_ENTRIES` were parsed with bare `int()` without try/except, unlike the TTL parsing which already had proper error handling. Added matching try/except blocks so invalid values fall back to defaults gracefully.

3. **Overly aggressive URL normalization** — The `tracking_prefixes` set included `ref` and `source`, which are legitimate query parameters on many sites (e.g., GitHub's `?ref=main`, API endpoints with `?source=internal`). Removed these two while keeping the actual tracking params (UTM, fbclid, gclid, etc.).

### Code quality improvements

4. **UPSERT pattern** — Replaced the SELECT → UPDATE/INSERT pattern in `set()` with a single `INSERT ... ON CONFLICT(url_hash) DO UPDATE` statement. Simpler, atomic, and explicitly preserves `access_count` on updates.

5. **Style cleanups** — Removed the `sys.path.insert` hack in `test_firecrawl_cache.py` (not needed with `pip install -e .`), removed the deprecated `test_cache_size_limits` test (already superseded by `test_cache_max_size_enforcement`), and fixed trailing whitespace in `preprocess.py`.

### What we kept as-is

The `content_hash` column is computed and stored but never read back. We left it in place since it's harmless and could be useful for future content-change detection.

All fixes were done TDD-style with 4 new regression tests. Final result: **147/147 tests passing**.

Thanks for the thorough implementation — the cache architecture, the CLI commands, and the test coverage are all well done. Really nice contribution! 👏
